examples/sampling/graphbolt/temporal_link_prediction.py

```diff
@@ -0,0 +1,356 @@
+main
+│
+├───> OnDiskDataset pre-processing
+│
+├───> Instantiate HeteroSAGE model
+│
+├───> train
+│     │
+│     ├───> Get graphbolt dataloader (HIGHLIGHT)
+│     │
+│     └───> Training loop
+│           │
+│           ├───> HeteroSAGE.forward
+│           │
+│           └───> Validation set evaluation
+│
+└───> Test set evaluation
+"""
+import argparse
+import os
+import time
+from functools import partial
+
+import dgl.graphbolt as gb
+import dgl.nn as dglnn
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import tqdm
+from dgl.data.utils import download, extract_archive
+
+
+TIMESTAMP_FEATURE_NAME = "__timestamp__"
+node_feature_keys = {
+    "Product": ["categoryId"],
+    "Query": ["categoryId"],
+}
+
+target_type = ("Query", "Click", "Product")
+all_connected_types = [
+    target_type,
+    ("Product", "reverse_Click", "Query"),
+    ("Product", "reverse_QueryResult", "Query"),
+    ("Query", "QueryResult", "Product"),
+]
+
+
+class CategoricalEncoder(nn.Module):
+    def __init__(
+        self,
+        num_categories,
+        out_size,
+    ):
+        super().__init__()
+        self.embed = nn.Embedding(num_categories, out_size)
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        nn.init.xavier_uniform_(self.embed.weight)
+
+    def forward(self, input_feat: torch.Tensor):
+        return self.embed(input_feat.view(-1))
+
+
+class HeteroSAGE(nn.Module):
+    def __init__(self, in_size, hidden_size):
+        super().__init__()
+        self.layers = nn.ModuleList()
+        sizes = [in_size, hidden_size]
+        for size in sizes:
+            self.layers.append(
+                dglnn.HeteroGraphConv(
+                    {
+                        etype: dglnn.SAGEConv(
+                            size,
+                            hidden_size,
+                            "mean",
+                        )
+                        for etype in all_connected_types
+                    },
+                    aggregate="sum",
+                )
+            )
+        self.hidden_size = hidden_size
+        self.predictor = nn.Sequential(
+            nn.Linear(hidden_size, hidden_size),
+            nn.ReLU(),
+            nn.Linear(hidden_size, hidden_size),
+            nn.ReLU(),
+            nn.Linear(hidden_size, 1),
+        )
+
+    def forward(self, blocks, X_node_dict):
+        H_node_dict = X_node_dict
+        for layer_idx, (layer, block) in enumerate(zip(self.layers, blocks)):
+            H_node_dict = layer(block, H_node_dict)
+            is_last_layer = layer_idx == len(self.layers) - 1
+            if not is_last_layer:
+                H_node_dict = {
+                    ntype: F.relu(H) for ntype, H in H_node_dict.items()
+                }
+        return H_node_dict
+
+
+def create_dataloader(args, graph, features, itemset, is_train=True):
+    """Get a GraphBolt version of a dataloader for link prediction tasks. This
```

Comment:
Remove graphbolt related comments, this example is not for teaching people using GB, and should focus on Temporal related code.